File: neetcode150/10_backtracking/0051_n_queens.py
"""LeetCode 51: N-Queens - Hard"""
import time, tracemalloc
from typing import List

# Variation 1: Backtracking with sets (cols, diagonals)
class Solution_v1:
    def solveNQueens(self, n: int) -> List[List[str]]:
        cols, pos_diag, neg_diag = set(), set(), set()
        res = []; board = [["." for _ in range(n)] for _ in range(n)]
        def bt(r):
            if r == n:
                res.append(["".join(row) for row in board]); return
            for c in range(n):
                if c in cols or (r + c) in pos_diag or (r - c) in neg_diag: continue
                cols.add(c); pos_diag.add(r+c); neg_diag.add(r-c); board[r][c] = "Q"
                bt(r + 1)
                cols.remove(c); pos_diag.remove(r+c); neg_diag.remove(r-c); board[r][c] = "."
        bt(0); return res

# Variation 2: Bitmask backtracking (compact representation)
class Solution_v2:
    def solveNQueens(self, n: int) -> List[List[str]]:
        res = []; queens = []
        def bt(r, cols, pos_d, neg_d):
            if r == n:
                board = []
                for qc in queens:
                    row = "." * qc + "Q" + "." * (n - qc - 1); board.append(row)
                res.append(board); return
            available = (~(cols | pos_d | neg_d)) & ((1 << n) - 1)
            while available:
                bit = available & (-available)  # lowest set bit
                available &= available - 1
                c = bit.bit_length() - 1
                queens.append(c)
                bt(r + 1, cols | bit, (pos_d | bit) >> 1, (neg_d | bit) << 1)
                queens.pop()
        bt(0, 0, 0, 0); return res

if __name__ == "__main__":
    n = 4
    for Vi, Cls in enumerate([Solution_v1, Solution_v2], 1):
        tracemalloc.start(); t0 = time.perf_counter()
        r = Cls().solveNQueens(n)
        t1 = time.perf_counter(); mem = tracemalloc.get_traced_memory()[1]; tracemalloc.stop()
        print(f"var{Vi}: {len(r)} solutions, mem={mem}bytes, time={(t1-t0)*1e6:.2f}us")
# var1 mem = {} and time = {}
# var2 mem = {} and time = {}

File: neetcode150/10_backtracking/0051_n_queens.cpp
#include <bits/stdc++.h>
#include <chrono>
using namespace std;
/* LeetCode 51: N-Queens - Hard */
int n_val;
set<int> cols_s, pd_s, nd_s;
vector<int> queens_s;
vector<vector<string>> res_s;
void bt(int r){
    if(r==n_val){ vector<string> board; for(int qc:queens_s){ string row(n_val,'.'); row[qc]='Q'; board.push_back(row); } res_s.push_back(board); return; }
    for(int c=0;c<n_val;c++){
        if(cols_s.count(c)||pd_s.count(r+c)||nd_s.count(r-c)) continue;
        cols_s.insert(c); pd_s.insert(r+c); nd_s.insert(r-c); queens_s.push_back(c);
        bt(r+1);
        cols_s.erase(c); pd_s.erase(r+c); nd_s.erase(r-c); queens_s.pop_back();
    }
}
int main(){
    n_val=4;
    auto s=chrono::high_resolution_clock::now();
    bt(0); cout<<"var1: "<<res_s.size()<<" solutions"<<endl;
    auto e=chrono::high_resolution_clock::now();
    cout<<"var1 time="<<chrono::duration_cast<chrono::nanoseconds>(e-s).count()/1000.0<<"us"<<endl;
}
// var1 mem = {} and time = {}
